File: amazon/103_binary_tree_zigzag_level_order_traversal.scala
// Given the root of a binary tree, return the zigzag level order 
// traversal of its nodes  values. 
// (i.e., from left to right, then right to left for the next 
// level and alternate between).

// Input: root = [3,9,20,null,null,15,7]
// Output: [[3],[20,9],[15,7]]
// Example 2:

// Input: root = [1]
// Output: [[1]]
// Example 3:

// Input: root = []
// Output: []

/**
 * Definition for a binary tree node.
 * class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
 *   var value: Int = _value
 *   var left: TreeNode = _left
 *   var right: TreeNode = _right
 * }
 */
object Solution {
    def zigzagLevelOrder(root: TreeNode): List[List[Int]] = {
        if(root == null) return List.empty[List[Int]]

        def bfs(level: List[TreeNode], direction: Boolean): List[List[Int]]= {
            if(level.isEmpty) return Nil

            val nextLevel = level.flatMap(node=>List(Option(node.left),Option(node.right)).flatten)
            val values = if(direction)level.map(_.value)else level.map(_.value).reverse

            values :: bfs(nextLevel, !direction)
        }
        bfs(List(root), true)
    }
}
 

File: amazon/103_binary_tree_zigzag_level_order_traversal.py


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root is None:
            return[]

        result = []
        node_queue = deque([root])
        left_to_right = True

        while node queue:
            level_size = len(node_queue)
            current_level = []

            for _ in range(level_size):
                current_node = node_queue.popleft()
                current_level.append(current_node.val)

                if current_node.left:
                    node_queue.append(current_node.left)
                if current_node.right:
                    node_queue.append(current_node.right)

            if left_to_right:
                result.append(current_level[:])
            else:
                result.append(current_level[::-1])

            left_to_right = not left_to_right

        return result
